web-service/helpers.py

```diff
@@ -132,42 +130,48 @@ def SortByDistanceCmp(a, b):
 
 ################################################################################
 
-def FetchAndStoreUrl(siteInfo, url):
+def GetSiteInfoForUrl(url, rssi=None, txpower=None, force_update=False):
+    siteInfo = models.SiteInformation.get_by_id(url)
+
+    if force_update or siteInfo is None:
+        siteInfo = FetchAndStoreUrl(siteInfo, url, rssi, txpower, force_update)
+
+    return siteInfo
+
+################################################################################
+
+def FetchAndStoreUrl(siteInfo, url, rssi=None, txpower=None, force_update=False):
     # Index the page
     try:
-        result = urlfetch.fetch(url, validate_certificate = True)
+        headers = {}
+        if rssi is not None and txpower is not None:
+            headers['X-PhysicalWeb-Rssi'] = rssi
+            headers['X-PhysicalWeb-TxPower'] = txpower
+
+        result = urlfetch.fetch(url,
+                                follow_redirects=False,
+                                validate_certificate=True,
+                                headers=headers)
     except:
         return StoreInvalidUrl(siteInfo, url)
 
-    if result.status_code == 200:
+    if result.status_code == 200: # OK
         encoding = GetContentEncoding(result.content)
-        final_url = GetExpandedURL(url)
-        real_final_url = result.final_url
-        if real_final_url is None:
-            real_final_url = final_url
-        return StoreUrl(siteInfo, url, final_url, real_final_url, result.content, encoding)
+        assert result.final_url is None
+        # TODO: Use the cache-content headers for storeUrl!
+        return StoreUrl(siteInfo, url, result.content, encoding)
+    elif result.status_code == 204: # No Content
+        # TODO: What do we return?  we want to filter this result out
+        return None
+    elif result.status_code in [301, 302, 303, 307, 308]: # Moved Permanently, Found, See Other, Temporary Redirect, Permanent Redirect
+        final_url = result.headers['location']
+        # TODO: Most redirects should not be cached, but we should still check!
```

Comment:
Note: Moved permanently usually have a cache-content time of 0, so its rare in practice that these should actually be cached.